### DISCUSSION TITLE
AdoProvider with AzureAD authentication

### DISCUSSION THREAD

**Discussion by Cisien** [quartznet/quartznet#1036] [Q&A] [answered]
I'm attempting to find a way to authenticate to Azure SQL using AAD authentication instead of having credentials in the connection string. 

Something like 
```cs
var connection = new SqlConnection(connectionString);
var tokenProvider = new AzureServiceTokenProvider();
connection.AccessToken = await tokenProvider.GetAccessTokenAsync("https://database.windows.net/");
```

I suspect this is something i'm going to need write some custom code to facilitate, but i'm having trouble finding the starting point.

**Comment by lahma** [comment, marked as answer]
I think that you can get away with something like this (pseudocode mostly).. *Edited after morning coffee kicked in*

```csharp
    public class AzureAadProvider : DbProvider
    {
        // here passing known supported DB provider name to base class
        public AzureAadProvider(string connectionString)
            : base("SqlServer", connectionString)
        {
        }

        public override DbConnection CreateConnection()
        {
            var connection = base.CreateConnection();
            var tokenProvider = new AzureServiceTokenProvider();
            connection.AccessToken = await tokenProvider.GetAccessTokenAsync("https://database.windows.net/");
            return connection;
        }
    }

....

    // before quartz initialization
    var provider = new AzureAadProvider("connection string here");
    Quartz.Util.DBConnectionManager.Instance.AddConnectionProvider("aadDataSource", provider);
```

Just don't configure any data sources in the configuration file. Just reference this manually registered via `quartz.jobStore.dataSource = aadDataSource`.

So not as straightforward as you might hope, but I hope this helps you to get started.

**Reply by Cisien** [reply]
Thanks for the starting point!

**Reply by Cisien** [reply]
I was able to get past the authentication hurdle. For completeness, here's the code that I was able to get working
```cs
    public class AadEnabledDbProvider : DbProvider
    {
        private readonly AzureServiceTokenProvider _tokenProvider = new AzureServiceTokenProvider();

        public AadEnabledDbProvider(string connectionString) : base("SqlServer", connectionString)
        {
        }

        public override DbConnection CreateConnection()
        {
            var connection = base.CreateConnection();
            if(connection is not SqlConnection sqlConnection)
            {
                throw new NotSupportedException("AAD Auth attempted on a DbConnection that doesn't support token authentication.");
            }

            sqlConnection.AccessToken = _tokenProvider.GetAccessTokenAsync("https://database.windows.net/").GetAwaiter().GetResult();
            return sqlConnection;
        }
    }
```

**Reply by Diaskhan** [reply]
oh man u make my day

**Reply by eluchsinger** [reply]
This solution is using a sync method to fetch a token asynchronously. Any alternatives?

**Comment by Diaskhan** [comment]
There is some update.
in 3.12 some things breaks.
https://github.com/quartznet/quartznet/releases/tag/v3.12.0

need add this 
`services.AddSingleton<IDbConnectionManager>(DBConnectionManager.Instance);`
before u gonna use 
Quartz.Util.DBConnectionManager.Instance.AddConnectionProvider("aadDataSource", provider);

**Comment by Mathyn** [comment]
Anyone with any experience setting this up to work with Azure Postgres?
